### PR TITLE
Flexible API for defining connection configuration

### PR DESCRIPTION
**Pull request by bilderbuchi**
Flexible API for defining connection configuration as per comment https://github.com/pymeasure/pymeasure/pull/536#discussion_r763372161

Connection configuration is now passed fully as kwargs to Instrument.init, and sorted out within VISAAdapter.
Rebased from an earlier iteration, will be completed and documented before review.

### CODE REVIEW

File: pymeasure/adapters/visa.py [hunk 1]
```diff
@@ -71,34 +76,35 @@ def has_supported_version():
...
+        :param visa_library: Optional pyvisa visa_library specifier for creating the resource manager.
         """
         # Get VISA interface type associated with resource_name
-        if isinstance(resource_name, int):
-            resource_name = "GPIB0::%d::INSTR" % resource_name
-        elif isinstance(resource_name, str):
-            pass
-        else:
-            # User defined adapter, do not touch anything
-            return
```
Comment (bilderbuchi) [inline pymeasure/adapters/visa.py:90]: this can be removed because it duplicates what already happens in VISAAdapter right before this method is called.

**Review by msmttchr — COMMENTED** [review]
@bilderbuchi, I am sending few comments on the proposed changes. Sharing my gut feeling: we are back at the `safeKeyword` approach.

File: pymeasure/adapters/visa.py [hunk 2]
```diff
@@ -71,34 +76,35 @@ def has_supported_version():
...
+            resource_mgr = pyvisa.ResourceManager(**visa_kwargs)
+
+        if_type = resource_mgr.resource_info(resource_name).interface_type
+
+        for key in list(received_kwargs):
+            # Remove all interface-specific kwargs:
+            if key in pyvisa.constants.InterfaceType.__members__:
+                if getattr(pyvisa.constants.InterfaceType, key) is if_type:
+                    # For the expected interface, dump contents into received_kwargs first
+                    received_kwargs.update(received_kwargs[key])
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/adapters/visa.py:105]: Just a quick remark, will review in detail later: can you propose (e.g. as a comment here) test cases for any features you are missing? That would help finding a good resolution.

File: pymeasure/adapters/visa.py [hunk 3]
```diff
@@ -71,34 +76,35 @@ def has_supported_version():
             return False
 
     @staticmethod
-    def setdefaults(resource_name, interface_type, user_kwargs, **kwargs):
-        """ Change user_kwargs to add defaults specified by user for the selected interface type
+    def setdefaults(resource_name, received_kwargs, resource_mgr=None, visa_library=None):
+        """ Remove invalid entries from the received keyword arguments.
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/adapters/visa.py:80]: Good point, I already have a solution in my head.

File: pymeasure/adapters/visa.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/adapters/visa.py:105]: Nevermind, I'm on it.

[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/adapters/visa.py:105]: > In addition to allow the test test_serial_custom_baud_rate_is_set to pass a specific parameter needs to be added in the instrument interface.

I'm sorry, I don't understand, can you unpack that a bit?

[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/adapters/visa.py]: I think I resolved this now. Do you agree?

File: pymeasure/adapters/visa.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/adapters/visa.py:80]: I think I resolved this now. Do you agree?

File: pymeasure/adapters/visa.py [hunk 4]
```diff
@@ -57,6 +57,11 @@ def __init__(self, resource_name, visa_library='', preprocess_reply=None, **kwar
             resource_name = "GPIB0::%d::INSTR" % resource_name
         self.resource_name = resource_name
         self.manager = pyvisa.ResourceManager(visa_library)
+        # We can make that one non-static and use self.resource_name and self.manager,
+        # then it would get smaller.
+        # A better name would be "clean_arguments"
```
Comment (bilderbuchi) [inline pymeasure/adapters/visa.py:62]: @msmttchr what do you prefer? Mind if I make this a class method, maybe _private to indicate internal use only, or do you see this being useful outside the class somewhere?

**Comment by msmttchr** [comment]
Sorry for being cryptic. I meant that to have an overridable default for
baud_rate, you were forced to add a specific parameter with default value
and then use this parameter as argument of the baud_rate parameter in the
instrument definition.

In other words, the customer passes a parameter which have a default (this
is just python), this parameter is passed to  a new parameter called aslr
via dictionary and finally it is set via the setdefault call. In my
opinion, this is a bit clunky and not straightforward, but of course this
is a matter of taste.

Il Mer 8 Dic 2021, 20:59 Christoph Buchner ***@***.***> ha
scritto:

**Comment by bilderbuchi** [comment]
> Sorry for being cryptic. I meant that to have an overridable default for
baud_rate, you were forced to add a specific parameter with default value
and then use this parameter as argument of the baud_rate parameter in the
instrument definition.

Ah thanks. I didn't notice that adding a specific parameter with default value is necessary. In any case, now it is not necessary anymore. :-)

File: pymeasure/adapters/visa.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (msmttchr) [reply pymeasure/adapters/visa.py]: Yes I agree

File: pymeasure/adapters/visa.py (hunk 4 above)
Comment (msmttchr) [reply pymeasure/adapters/visa.py:62]: I think that, according to your approach, this method doesn't make sense anymore, even as internal method. I would remove it altogether and replace it with inline code where is needed.

**Comment by bilderbuchi** [comment]
@CasperSchippers @LongnoseRob please review this implementation a way to set/override connection defaults.

We defined tests that nail down the desired behaviour for users and instrument implementors.
I added a 7-line (without comments/docstrings) portion to VISAAdapter
I added and updated docs both from the user side ("how do I tweak connection settings for an existing instrument") and the implementor side ("how do I set up an instrument with proper connection defaults, both for single- and multiinterface devices).
Finally, I converted any instruments that I could find to the new best practice.

I will be happy to get any feedback on this. Did we miss a use case? Is the documentation clear? Can the docs be improved? Can the _implementation_ be improved?

**Comment by LongnoseRob** [comment]
> @CasperSchippers @LongnoseRob please review this implementation a way to set/override connection defaults.
> 
> We defined tests that nail down the desired behaviour for users and instrument implementors. I added a 7-line (without comments/docstrings) portion to VISAAdapter I added and updated docs both from the user side ("how do I tweak connection settings for an existing instrument") and the implementor side ("how do I set up an instrument with proper connection defaults, both for single- and multiinterface devices). Finally, I converted any instruments that I could find to the new best practice.
> 
> I will be happy to get any feedback on this. Did we miss a use case? Is the documentation clear? Can the docs be improved? Can the _implementation_ be improved?

From just reading the changes it looks fine for me.
For testing it with the HP3478A: I might be able to do that later today or in the next days.

BTW: for the timeout on GPIB with pyvisa-py as backend (and maybe other backends aswell), the timeout cannot be set to arbitrary values. (Something that I learnt yesterday)

**Comment by bilderbuchi** [comment]
> BTW: for the timeout on GPIB with pyvisa-py as backend (and maybe other backends aswell), the timeout cannot be set to arbitrary values. (Something that I learnt yesterday)

https://github.com/pyvisa/pyvisa-py/blob/40d439627aa6fd57f36093073c1e661f108cab17/pyvisa_py/gpib.py#L140

hm, this seems to be a limitation of a linux gpib library, and seems to be restricted to pyvisa-py and gpib, only. I can't find any indication otherwise in the pyvisa or pyvisa-py codebase.

**Comment by LongnoseRob** [comment]
> > BTW: for the timeout on GPIB with pyvisa-py as backend (and maybe other backends aswell), the timeout cannot be set to arbitrary values. (Something that I learnt yesterday)
> 
> hm, this seems to be a [limitation of a linux gpib library](https://github.com/pyvisa/pyvisa-py/blob/40d439627aa6fd57f36093073c1e661f108cab17/pyvisa_py/gpib.py#L140), and seems to be restricted to pyvisa-py and gpib, only. I can't find any indication otherwise in the pyvisa or pyvisa-py codebase.

Yes, that is the same source I was looking at...
### Back to the topic:
I tested the  changes with [my verification file](https://github.com/LongnoseRob/misc/blob/3d26061585a6e8024044f873bbe4e1d3bdae26d3/3478_verifications.py) with shorted inputs (2 wire method)

output of the run  is [here](https://gist.github.com/LongnoseRob/d45dc1e9fa1b6f6e74cd7f6ccf7a77d2)
_(sorry it is a not very elaborate reporting test file)_

--> looks good to me and **"go"** for the merge from me

**Review by CasperSchippers — APPROVED** [review]
Nice job (to all, @bilderbuchi, @msmttchr, @LongnoseRob), and sorry for not contributing so far to this discussion.

This implementation looks pretty good to me; I like how it enables you to simply use keyword arguments to fully define the default communication parameters for all possible communication protocols, and similarly only need keyword arguments to overrule the defaults.

I didn't spot any issues or improvements. So for me it seems good to go.

PS I tested it with the Oxford instrument ITC503 (the only device that has been changed in this PR that I have access to) and it works perfectly fine as far as I can tell.

File: pymeasure/instruments/instrument.py [hunk 5]
```diff
@@ -35,14 +35,35 @@
...
 class Instrument(object):
-    """ This provides the base class for all Instruments, which is
-    independent of the particular Adapter used to connect for
-    communication to the instrument. It provides basic SCPI commands
-    by default, but can be toggled with :code:`includeSCPI`.
+    """ The base class for all Instrument definitions.
 
-    :param adapter: An :class:`Adapter<pymeasure.adapters.Adapter>` object
-    :param name: A string name
+    It makes use of an array of :py:class:`~pymeasure.adapters.Adapter` classes for communication
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/instruments/instrument.py:40]: Thanks, done!

File: pymeasure/adapters/visa.py [hunk 6]
```diff
@@ -57,6 +81,20 @@ def __init__(self, resource_name, visa_library='', preprocess_reply=None, **kwar
             resource_name = "GPIB0::%d::INSTR" % resource_name
         self.resource_name = resource_name
         self.manager = pyvisa.ResourceManager(visa_library)
+
+        # Clean up kwargs considering the interface type matching resource_name
+        if_type = self.manager.resource_info(self.resource_name).interface_type
+        for key in list(kwargs):
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/adapters/visa.py:87]: Indeed, I agree.

File: pymeasure/adapters/visa.py [hunk 7]
```diff
@@ -57,6 +81,20 @@ def __init__(self, resource_name, visa_library='', preprocess_reply=None, **kwar
...
+            # Remove all interface-specific kwargs:
+            if key in pyvisa.constants.InterfaceType.__members__:
+                if getattr(pyvisa.constants.InterfaceType, key) is if_type:
+                    # For the present interface, dump contents into kwargs first if they are not
+                    # present already. This way, it is possible to override default values with
+                    # kwargs passed to Instrument.__init__()
+                    for k, v in kwargs[key].items():
+                        kwargs.setdefault(k, v)
+                del kwargs[key]
```
[the remark this answers is not in the document]
Comment (bilderbuchi) [reply pymeasure/adapters/visa.py:96]: Yes it does, I added a test for this nonetheless

**Comment by bilderbuchi** [comment]
Thanks for the review, I'll merge this!

**Pull request merged**